### ISSUE TITLE
[no-constant-condition] all literal should be treated like boolean literal

### ISSUE THREAD

**Issue by Zzzen** [eslint/eslint#13238]
**What rule do you want to change?**
no-constant-condition

**Does this change cause the rule to produce more or fewer warnings?**
More warnings

**How will the change be implemented? (New option, new default behavior, etc.)?**
New option

**Please provide some example code that this change will affect:**

```js
// Unexpected constant condition. eslintno-constant-condition
if (true || this.loginpage) {
    console.log('1');
}

// no error reported
if (1 || this.loginpage) {
    console.log('1');
}
```

**What does the rule currently do for this code?**
Only the first `if` statement is reported

**What will the rule do after it's changed?**
The second `if` statement will be reported too.

**Are you willing to submit a pull request to implement this change?**
Yes

**Comment by yeonjuan** [comment]
[Online Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tY29uc3RhbnQtY29uZGl0aW9uOiBcImVycm9yXCIqL1xuXG5pZiAodGhpcy5sb2dpbnBhZ2UgfHx0cnVlKSB7fSAvLyBFcnJvclxuaWYgKHRoaXMubG9naW5wYWdlIHx8IDEpIHt9IC8vIEVycm9yXG5pZiAodGhpcy5sb2dpbnBhZ2UgfHwgJzEnKSB7fSAvLyBFcnJvclxuXG5pZiAodHJ1ZSB8fCB0aGlzLmxvZ2lucGFnZSkge30gLy8gRXJyb3JcblxuaWYgKDEgfHwgdGhpcy5sb2dpbnBhZ2UpIHt9IC8vIE5vIGVycm9yXG5pZiAoJzEnIHx8IHRoaXMubG9naW5wYWdlKSB7fSAvLyBObyBFcnJvclxuIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoxMSwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJzb3J0LWltcG9ydHMiOjJ9LCJlbnYiOnsiZXMyMDIwIjp0cnVlfX19)
```js
/*eslint no-constant-condition: "error"*/

if (this.loginpage ||true) {} // error
if (this.loginpage || 1) {} // error
if (this.loginpage || '1') {} // error

if (true || this.loginpage) {} // error

if (1 || this.loginpage) {} // no error
if ('1' || this.loginpage) {} // no error
```
Hi, @Zzzen. :). It seems a bug to me.
I think the truthy literal on the left side of `||` should be reported.

> How will the change be implemented? (New option, new default behavior, etc.)?
New option

so, maybe, it should be fixed rather than adding new options?
Let’s hear from others.

**Comment by anikethsaha** [comment]
I agree with above comment. 

```js
if (1 || this.loginpage) {} // this should be an error cause its always truthy
if ('1' || this.loginpage) {} // this should be an error , always truthy
```

**Comment by Zzzen** [comment]
Because I'm not sure what's the rationale behind these lines 😂 

https://github.com/eslint/eslint/blob/fdfa364c1b0d88689d02cbe8ae848a3e323209aa/tests/lib/rules/no-constant-condition.js#L76

```js
"if('str' || a){}",
"if('str1' && a){}",
"if(a && 'str'){}",
"if('str' || abc==='str'){}",
```

**Comment by Zzzen** [comment]
Do we have any rules for `if (false || a){}`? It's likely a mistake ( just like an unused expression ), although not a constant condition.

**Comment by anikethsaha** [comment]
I think it should be reported as well with an unused condition as `if (false || a){}` is equal to `if (a){}`. It can be done in this rule itself instead of a separate rule.

**Comment by Zzzen** [comment]
It would be a big breaking change.

**Comment by mdjermanovic** [comment]
> Do we have any rules for `if (false || a){}`? It's likely a mistake ( just like an unused expression ), although not a constant condition.

We don't have such a rule. It might be worth exploring whether an option in `no-constant-condition` to treat LHS of a logical operator as condition makes sense.

**Comment by mdjermanovic** [comment]
I believe the only reason to not report `1 || foo` as a constant condition in `if (1 || foo);` would be to avoid overlapping with another rule (or an option in this rule) that would report `1` in `1 || foo`.

In that case, `true || foo` in `if (true || foo);` shouldn't be reported as well.

`foo || 1` is different because it can be a valid code, depending on the context.

Even if we add such a rule or option at some point, I don't think that the mentioned overlapping would be a big problem, so I'd vote to accept this issue as a bug and fix it in a semver-minor release.

In addition to `||`, the actual behavior with `&&` also seems to be inconsistent:

```js
/* eslint no-constant-condition: error */

if (false && a); // error

if (void null && a); // error!

if (null && a); // no error
```
